cli/internal/run/run.go

```diff
@@ -311,161 +274,84 @@ func (r *run) runOperation(ctx gocontext.Context, g *graph.CompleteGraph, rs *ru
-		if r.opts.runOpts.singlePackage {
-			r.base.UI.Output(fmt.Sprintf("%s %s", ui.Dim("• Running"), ui.Dim(ui.Bold(strings.Join(rs.Targets, ", ")))))
-		} else {
-			r.base.UI.Output(fmt.Sprintf(ui.Dim("• Packages in scope: %v"), strings.Join(packagesInScope, ", ")))
-			r.base.UI.Output(fmt.Sprintf("%s %s %s", ui.Dim("• Running"), ui.Dim(ui.Bold(strings.Join(rs.Targets, ", "))), ui.Dim(fmt.Sprintf("in %v packages", rs.FilteredPkgs.Len()))))
-		}
-		return r.executeTasks(ctx, g, rs, engine, packageManager, tracker, startAt)
+		return GraphRun(ctx, rs, engine, r.base)
 	}
-	return nil
-}
 
-func renderDryRunSinglePackageJSON(tasksRun []hashedTask) (string, error) {
-	singlePackageTasks := make([]hashedSinglePackageTask, len(tasksRun))
-	for i, ht := range tasksRun {
-		singlePackageTasks[i] = ht.toSinglePackageTask()
-	}
-	dryRun := &struct {
-		Tasks []hashedSinglePackageTask `json:"tasks"`
-	}{singlePackageTasks}
-	bytes, err := json.MarshalIndent(dryRun, "", "  ")
-	if err != nil {
-		return "", errors.Wrap(err, "failed to render JSON")
-	}
-	return string(bytes), nil
-}
+	packagesInScope := rs.FilteredPkgs.UnsafeListOfStrings()
+	sort.Strings(packagesInScope)
+	// Initiate analytics and cache
+	analyticsClient := r.initAnalyticsClient(ctx)
+	defer analyticsClient.CloseWithTimeout(50 * time.Millisecond)
+	turboCache, err := r.initCache(ctx, rs, analyticsClient)
 
-func renderDryRunFullJSON(tasksRun []hashedTask, packagesInScope []string) (string, error) {
-	dryRun := &struct {
-		Packages []string     `json:"packages"`
-		Tasks    []hashedTask `json:"tasks"`
-	}{
-		Packages: packagesInScope,
-		Tasks:    tasksRun,
-	}
-	bytes, err := json.MarshalIndent(dryRun, "", "  ")
 	if err != nil {
-		return "", errors.Wrap(err, "failed to render JSON")
-	}
-	return string(bytes), nil
-}
-
-func displayDryTextRun(ui cli.Ui, tasksRun []hashedTask, packagesInScope []string, packageInfos map[interface{}]*fs.PackageJSON, isSinglePackage bool) error {
-	if !isSinglePackage {
-		ui.Output("")
-		ui.Info(util.Sprintf("${CYAN}${BOLD}Packages in Scope${RESET}"))
-		p := tabwriter.NewWriter(os.Stdout, 0, 0, 1, ' ', 0)
-		fmt.Fprintln(p, "Name\tPath\t")
-		for _, pkg := range packagesInScope {
-			fmt.Fprintf(p, "%s\t%s\t\n", pkg, packageInfos[pkg].Dir)
-		}
-		if err := p.Flush(); err != nil {
-			return err
+		if errors.Is(err, cache.ErrNoCachesEnabled) {
+			r.base.UI.Warn("No caches are enabled. You can try \"turbo login\", \"turbo link\", or ensuring you are not passing --remote-only to enable caching")
+		} else {
+			return errors.Wrap(err, "failed to set up caching")
 		}
 	}
 
-	ui.Output("")
-	ui.Info(util.Sprintf("${CYAN}${BOLD}Tasks to Run${RESET}"))
+	// Dry Run
+	if rs.Opts.runOpts.dryRun {
+		return DryRun(
+			ctx,
+			g,
+			rs,
+			engine,
+			tracker,
+			turboCache,
+			packagesInScope,
+			r.base,
+		)
+	}
+
+	// Regular run
+
+	return RealRun(
+		ctx,
+		g,
+		rs,
+		engine,
+		tracker,
+		turboCache,
+		packagesInScope,
+		r.base,
```

Comment:
Maybe put the shared args into a helper struct?

Reply:
I wanted to leave it super explicit and long and arduous right now, because I think we can limit the scope of things that get passed all over the place. I think a better architectural pattern can emerge if we go further down, but if we hide it in another struct, it won't be as obvious. One way I think about it is... "what would the struct be called". If I don't have a good answer (and I don't), I shouldn't do it quite yet.